File: push/src/main/java/top/trumeet/mipushframework/wizard/PermissionInfoFactory.java
package top.trumeet.mipushframework.wizard;

import android.content.Context;
import android.content.Intent;

public class PermissionInfoFactory {
    private static final String intentKey = PermissionInfo.class.getSimpleName();

    public static PermissionInfo create(String className, Context context) {
        if (UsageStatsPermissionInfo.class.getSimpleName().equals(className)) {
            return new UsageStatsPermissionInfo(context);
        }
        return null;
    }

    public static Intent bindPermissionInfo(
            Intent intent, Class<? extends PermissionInfo> permissionInfoClass) {
        return intent.putExtra(intentKey, permissionInfoClass.getSimpleName());
    }

    public static PermissionInfo createFrom(Intent intent, Context context) {
        String permissionInfoClass = intent.getStringExtra(intentKey);
        return create(permissionInfoClass, context);
    }

}

File: push/src/main/java/io/github/magisk317/mipush/feature/wizard/permission/PermissionInfoFactory.kt
package io.github.magisk317.mipush.feature.wizard.permission

import android.content.Context
import android.content.Intent

object PermissionInfoFactory {
    private val intentKey = PermissionInfo::class.java.simpleName

    @JvmStatic
    fun create(className: String?, context: Context): PermissionInfo? {
        if (className == null) {
            return null
        }
        return try {
            val clazz = Class.forName(className)
            val constructor = clazz.getConstructor(Context::class.java)
            constructor.newInstance(context) as PermissionInfo
        } catch (e: Exception) {
            io.github.aakira.napier.Napier.w("PermissionInfoFactory: failed to create ${clazz.simpleName}", e, tag = "PermissionInfoFactory")
            null
        }
    }

    @JvmStatic
    fun bindPermissionInfo(
        intent: Intent,
        permissionInfoClass: Class<out PermissionInfo>
    ): Intent {
        return intent.putExtra(intentKey, permissionInfoClass.canonicalName)
    }

    @JvmStatic
    fun createFrom(intent: Intent, context: Context): PermissionInfo? {
        val permissionInfoClass = intent.getStringExtra(intentKey)
        return create(permissionInfoClass, context)
    }
}
